### PR TITLE
Add isnan and iszero

### LINKED ISSUE #7273: Add `isnan` and `iszero` for FLOAD and DOUBLE
### Is your feature request related to a problem or challenge?

In Datafusion, `+NaN` and` -NaN` are distinguished as well as `+0.0` and `-0.0`.
For example, `asin(10)` returns `NaN` but `'inf'::DOUBLE / 'inf'::DOUBLE` returns `-NaN`.
`3.0 * 0` returns `0.0` but `-3.0 * 0` returns `-0.0`.

So, users need to determine `NaN` and `0.0` like as follows.

```
SELECT ... WHERE x = 'NaN'::DOUBLE OR x = -'NaN'::DOUBLE;
```

To avoid such boilerplate code, I propose to add `isnan` and `iszero`.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sarutak**
## Which issue does this PR close?

Closes #7273 

## Rationale for this change
In Datafusion, `+NaN` and` -NaN` are distinguished as well as `+0.0` and `-0.0`.
For example, `asin(10)` returns `NaN` but `'inf'::DOUBLE / 'inf'::DOUBLE` returns `-NaN`.
`3.0 * 0` returns `0.0` but `-3.0 * 0` returns `-0.0`.

So, users need to determine `NaN` and `0.0` like as follows.

```
SELECT ... WHERE x = 'NaN'::DOUBLE OR x = -'NaN'::DOUBLE;
```

I think it's nice to avoid such boilerplate code.

## What changes are included in this PR?
This PR aims to add `isnan` and `iszero` which determine `NaN` and `0.0` ignoring their sign.

## Are these changes tested?
Added new tests.

## Are there any user-facing changes?
Yes but it doesn't break compatibility.

### CODE REVIEW

File: docs/source/user-guide/expressions.md [hunk 1]
```diff
@@ -94,6 +94,8 @@ expressions such as `col("a") + col("b")` to be used.
 | factorial(x)          | factorial                                         |
 | floor(x)              | nearest integer less than or equal to argument    |
 | gcd(x, y)             | greatest common divisor                           |
+| isnan(x)              | predicate determining whether NaN or not          |
+| iszero(x)             | predicate determining whether 0.0 or not          |
```
Comment (viirya) [inline docs/source/user-guide/expressions.md:98]:
```suggestion
| isnan(x)              | predicate determining whether NaN/-NaN or not          |
| iszero(x)             | predicate determining whether 0.0/-0.0 or not          |
```

File: datafusion/expr/src/built_in_function.rs [hunk 2]
```diff
@@ -1186,7 +1194,9 @@ impl BuiltinScalarFunction {
             | BuiltinScalarFunction::Sqrt
             | BuiltinScalarFunction::Tan
             | BuiltinScalarFunction::Tanh
-            | BuiltinScalarFunction::Cot => {
+            | BuiltinScalarFunction::Cot
+            | BuiltinScalarFunction::Isnan
+            | BuiltinScalarFunction::Iszero => {
```
Comment (viirya) [inline datafusion/expr/src/built_in_function.rs:1199]: I would consider `Isnan` and `Iszero` not math expressions as listed here. The reason to give f64 high priority is also not applied too for these two expressions, I think.

**Review by viirya — APPROVED** [review]
Looks good to me. They are also in Spark expressions too, although seems not all similar systems support these two expressions.

File: datafusion/expr/src/built_in_function.rs (hunk 2 above)
Comment (viirya) [reply datafusion/expr/src/built_in_function.rs:1199]: Based on above reason, maybe it is less confused to have them in a separate pattern.

**Review by Dandandan — APPROVED** [review]

File: datafusion/expr/src/expr_fn.rs [hunk 3]
```diff
@@ -811,6 +811,18 @@ scalar_expr!(CurrentDate, current_date, ,"returns current UTC date as a [`DataTy
 scalar_expr!(Now, now, ,"returns current timestamp in nanoseconds, using the same value for all instances of now() in same statement");
 scalar_expr!(CurrentTime, current_time, , "returns current UTC time as a [`DataType::Time64`] value");
 scalar_expr!(Nanvl, nanvl, x y, "returns x if x is not NaN otherwise returns y");
+scalar_expr!(
+    Isnan,
+    isnan,
+    num,
+    "returns true if a given number is +NaN or -NaN otherwise returns false"
```
[the remark this answers is not in the document]
Comment (tustvold) [reply datafusion/expr/src/expr_fn.rs:818]: Any value with an exponent field of all 1s is a Nan, and so there are 2^N distinct values of NaN, where N is the number of mantissa bits.

Within arrow-rs we follow the IEEE 754  total order predicate which establishes an ordering for NaNs, (and infinity, etc...)

File: datafusion/expr/src/built_in_function.rs (hunk 2 above)
Comment (sarutak) [reply datafusion/expr/src/built_in_function.rs]: Sorry for the late reply.
Yes, f64 doesn't need high priority.
I've changed it.

**Comment by sarutak** [comment]
This CI failure seems caused by #7270. I opened #7286 to fix it.

File: docs/source/user-guide/sql/scalar_functions.md [hunk 4]
```diff
@@ -283,6 +285,32 @@ gcd(expression_x, expression_y)
 - **expression_y**: Second numeric expression to operate on.
   Can be a constant, column, or function, and any combination of arithmetic operators.
 
+### `isnan`
```
Comment (alamb) [inline docs/source/user-guide/sql/scalar_functions.md:288]: ❤️

File: datafusion/sqllogictest/test_files/math.slt [hunk 5]
```diff
@@ -99,3 +99,15 @@ query RRR
 SELECT nanvl(asin(10), 1.0), nanvl(1.0, 2.0), nanvl(asin(10), asin(10))
 ----
 1 1 NaN
+
+# isnan
+query BBBB
```
Comment (liukun4515) [inline datafusion/sqllogictest/test_files/math.slt:104]: 👍

**Review by liukun4515 — APPROVED** [review]
LGTM, will merge it.

**Pull request merged**